### PR TITLE
compute global norm on device

### PR DESCRIPTION
**Pull request by BacharL**
Avoid host synchronization by keeping data on device

### CODE REVIEW

File: deepspeed/runtime/utils.py [hunk 1]
```diff
@@ -365,42 +365,52 @@ def clip_grad_norm_(parameters, max_norm, norm_type=2, mpu=None):
...
     max_norm = float(max_norm)
     norm_type = float(norm_type)
+    all_norms = []
     if norm_type == inf:
-        total_norm = max(p.grad.data.abs().max() for p in parameters)
-        total_norm_cuda = get_accelerator().FloatTensor([float(total_norm)])
+        for p in parameters:
+            all_norms.append(p.grad.data.abs().max().float())
+        total_norm = torch.stack(all_norms).max()
+        origin_device = total_norm.device.type
```
Comment (tjruwase) [inline deepspeed/runtime/utils.py:373]: Should this not be `total_norm.device`? Won't the current code lose the device index?

File: deepspeed/runtime/utils.py [hunk 2]
```diff
@@ -870,12 +880,15 @@ def get_global_norm_of_tensors(input_tensors, norm_type=2, mpu=None, use_graph=F
     assert all([torch.is_tensor(t) for t in input_tensors]), f'expected list of only tensors'
 
     norm_type = float(norm_type)
+    all_norms = []
     if norm_type == inf:
-        total_norm = max(t.data.abs().max() for t in input_tensors)
-        total_norm_cuda = get_accelerator().FloatTensor([float(total_norm)])
+        for t in input_tensors:
+            all_norms.append(t.data.abs().max().float())
+        total_norm_cuda = torch.stack(all_norms).max()
```
Comment (tjruwase) [inline deepspeed/runtime/utils.py:887]: Rename to make device-agnostic
```suggestion
        device_total_norm = torch.stack(all_norms).max()
```

File: deepspeed/runtime/utils.py (hunk 1 above)
Comment (BacharL) [reply deepspeed/runtime/utils.py]: fixed

File: deepspeed/runtime/utils.py (hunk 2 above)
Comment (BacharL) [reply deepspeed/runtime/utils.py]: done

**Comment by tjruwase** [comment]
@BacharL, thanks for the PR. Can you please help resolve the conflicts to unblock the CI? Thanks!

**Pull request merged**